Implement Audio cues on cell execution completed

MonadChains opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->
Fixes #165085
This PR implement an option in the settings `notebook.audioCuesEnabled` that enables audio cues on cell execution both in case of success and in case of failure.

roblourens commented on src/vs/workbench/contrib/notebook/common/notebookCommon.ts:
```diff
@@ -925,7 +925,8 @@ export const NotebookSetting = {
 	outputFontSize: 'notebook.outputFontSize',
-	outputFontFamily: 'notebook.outputFontFamily'
+	outputFontFamily: 'notebook.outputFontFamily',
+	audioCuesEnabled: 'notebook.audioCuesEnabled'
```
`notebook.audioCues.enabled`

roblourens reviewed (commented):
Thanks for tackling this! Looks fine, but it makes me wonder- if you click "Run all" (or run above/below, or select a range and run...) should it sound the audio cue for every cell individually, or just once at the end of the "batch"? I don't know

MonadChains commented:
Good point! I didn't think about it. I would lean more toward one sound per batch because it can be annoying when you have multiple small cells. On the other hand, having a sound for each cells could provide more informative feedback (I'm thinking of the possibility of counting the cues for knowing which cell is currently being executed). There is also the option of having the user deciding it with an additional setting.

meganrogge commented:
@roblourens we sort of encountered this with the build task audio cues and decided to just play one sound at a time instead of capturing the requests and playing them sequentially. https://github.com/microsoft/vscode/issues/164921

Do the cells get run in parallel? I imagine they don't necessarily finish in order, so it could be pretty confusing for the user if multiple sounds play... for example:

Cell 1
Cell 2
Cell 3

Cell 2 succeeds
Cell 1 fails
Cell 3 succeeds

Could sound misleading given the order the cells appear in is not consistent w the order in which the noise gets played.

MonadChains commented:
I've changed the behavior of the audio cues. Now the successful execution cue is played on the last cell scheduled, while the failed one always plays whenever a cell fails, since a failed execution interrupts the all the remaining scheduled ones.

roblourens commented on src/vs/workbench/contrib/notebook/browser/services/notebookExecutionStateServiceImpl.ts:
```diff
@@ -142,11 +147,25 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
-
+			exe.onDidComplete(onDidCompleteCallback));
+		this._cellListeners.set(cellUri, disposable);
+		this._lastScheduledCellUri = cellUri;
```
Not crazy about this approach- this service is global where we should track this per-notebook. And this tracks the last cell to start, and as Megan points out, in some notebooks cells can run in parallel and thus finish out of order. I think a better way to do this would be when a cell finishes running, check whether there are any other executing cells in the same notebook, if not, play the sound.

MonadChains replied:
Reverted to the old style + implemented the behaviour in case of multiple cell executed

roblourens commented on src/vs/workbench/contrib/notebook/browser/services/notebookExecutionStateServiceImpl.ts:
```diff
@@ -142,11 +147,25 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
+		let onDidCompleteCallback: (lastRunSuccess: boolean | undefined) => void;
+		if (audioCuesEnabled) {
+			const audioCuesCallback = this._getAudioCuesOnDidCompleteCallback(cellUri);
+			onDidCompleteCallback = (lastRunSuccess) => {
```
Feels like checking the setting outside of the callback is making this overly complicated. Checking it inside the callback might be simpler and also make it work if I change the setting while the cell is executing

MonadChains replied:
> make it work if I change the setting while the cell is executing
Actually what I wanted to achieve was the exact contrary. I thought that set the audio cue behaviour when the execution was scheduled would have been more consistent. But I see in terms value in terms of user experience for the behaviour you proposed.

roblourens commented:
Every other Audio Cues setting has values like this:

<img width="413" alt="image" src="https://user-images.githubusercontent.com/323878/201401047-6b97e2e4-b9c7-49b7-a0d9-741b541a4812.png">

please match that pattern. I'm not hearing your audio cue, that's probably because I don't have a screenreader enabled currently, but I hear the others when setting them to On

MonadChains commented:
I'm using the `audioCues.taskCompleted` cue for successful cells and `audioCues.taskFailed` for failed ones as proposed in the discussion of the issue. These audio cues must be enabled in the settings along with the newly added option in order to hear the sounds:
![Screenshot from 2022-11-11 20-45-07](https://user-images.githubusercontent.com/105170638/201418522-3673a7d3-2ca7-42a1-8d11-ae0702446436.png)


 The checks for the 2 cues are done at the level of the audioCueService:
https://github.com/microsoft/vscode/blob/b982536f83376e6266b3e0aaca6211699167b6cf/src/vs/workbench/contrib/audioCues/browser/audioCueService.ts#L42-L46
So to bypass it the options are:

- Modify the audioCueService
- Create new cues with the same sounds (don't like to duplicate stuff tbh)
- Call `playSound` directly in the callback (but this would remove a bit of the abstraction provided by the service)

There is also the option to scrap the new setting completely since cell execution can be considered a "task", and, probably users who turn on the options would likely want to have cues played for notebooks.

What do you think?I'm open to every option

roblourens commented:
A notebook cell is not a task. Looking at the code, there are more to audio cues than I thought- the notebook AudioCues should be registered with the others here: https://github.com/microsoft/vscode/blob/5608a44d25212260807bbf9f3cfa9ab27dd29960/src/vs/workbench/contrib/audioCues/browser/audioCueService.ts#L187

so that they should show up in the "List Audio Cues" command like the others. This isn't duplicating stuff- notebook cell execution is its own feature and should be listed and controlled independently.

MonadChains commented:
I've added the audioCues and removed the old setting. The settings for the cues will appear as `audioCues.notebookCellCompleted` and `audioCues.notebookCellFailed` in the settings. I've put the settings in the audioCues group instead of the notebook one because I've seen that the cues for other contribs such as the terminal or the debugger are in the audioCues group too.

roblourens reviewed (dismissed):
Love it, thanks!
